Generate sync code from async

dvarrazzo opened the pull request:
This massive refactoring is intended to unify sync and async code, and write it only once.

With this refactoring, most of the code that exist in both sync and async version must be only written as async; the sync version is generated automatically by a script.

The code converted automatically is:
- Connection, Cursor, Copy objects (other objects, such as Transaction, Pipeline, are not worth the refactoring, as the sync/async outer layer is very small).
- Connection pool classes, scheduler
- all tests except concurrency tests, which are too different between threads/async and cover unrelated cases.

The conversion script `async_to_sync.py` works at AST level: its main tasks are:

- convert async/await syntax (for, def, with, ...) to non-async
- Rename objects and attributes: `AsyncCursor` -> `Cursor`, `__aenter__` -> `__enter__`...
- Drop certain if branches which in the async code look like:
```python
if True:  # ASYNC
    ...
else:
    ...
```
in the sync code, only the `else` branch is left. `if False:  # ASYNC` is supported too. This is used to drop entire snippets of syntax which would be invalid in the sync code.

"acompat" modules have been introduced, exposing objects pairs (`aspawn()/spawn()`, `agather()/gather()`, `AEvent/Event`...) to unify certain subtle differences between tasks/threads (for instance `asyncio.Event.wait()` missing a `timeout` argument) and allow to write more similar sync/async code.

A wrapper script `convert_async_to_sync.sh` iterates over all the files to convert. It also offer a `--check` mode to use in CI in order to verify that the async and sync code are kept in sync.

Using the script was already useful to find a few little bugs which were fixed in 3.1.11:

- Fix possible lack of critical section guard in async `executemany()`.
- Fix missing pipeline fetch in async `scroll()`.

Keeping this out for comment for a few days before merging. @dlax, @asqui what do you think about this approach?

dlax commented:
> Keeping this out for comment for a few days before merging.

I'll review this as soon as I can, next week hopefully.

asqui commented on psycopg_pool/psycopg_pool/_acompat.py:
```diff
@@ -0,0 +1,162 @@
+"""
+Utilities to ease the differences between async and sync code.
+
+These object offer a similar interface between sync and async versions; the
+script async_to_sync.py will replace the async names with the sync names
+when generating the sync version.
+"""
```
This file seems to have a lot of overlap with `psycopg/psycopg/_acompat.py`
Is there any possibility to share code between the two packages?

dvarrazzo replied:
Yes, it's to keep them isolated from each other: there isn't a 1:1 version lock between pool and psycopg. 

If the pool used `_acompat` objects from psycopg it would become dependent on whatever version we would distribute those objects, and they would become a psycopg public interface that we shouldn't break.

We can afford to do that between psycopg and psycopg-c, because they have a 1:1 release pattern and each minor version is only expected to depend on the same minor version of the other package, but for the pool we decided on an release cycle independent from the main package.

asqui reviewed (approved):
Cool idea and implementation!

I am not keeping too close an eye on this project so have lost familiarity with the finer details of the code (as is probably evident from some of my comments/questions!)

One concern is that if we're auto-generating both code and the associated tests, is there something to catch the scenario where we fail to do so? (e.g. by accidentally not adding to the hard-coded list of targets in `convert_async_to_sync.sh`) I guess this will be easily visible in the diffs, but for a large-scale change like this it seems hard to be sure that nothing has gone missing somewhere?

dvarrazzo commented on psycopg_pool/psycopg_pool/base_null_pool.py:
```diff
@@ -0,0 +1,29 @@
+
+        return min_size, max_size
+
+    def _start_initial_tasks(self) -> None:
+        # Null pools don't have background tasks to fill connections
+        # or to grow/shrink.
+        return
+
+    def _maybe_grow_pool(self) -> None:
+        # null pools don't grow
+        pass
```
Right, thank you. `pass` is more conventional here.

dvarrazzo commented on psycopg_pool/psycopg_pool/pool_async.py:
```diff
@@ -142,23 +141,41 @@ def __init__(
         if open:
             self._open()
 
+    if False:  # ASYNC
```
No, it wouldn't be the same. I forgot to explain the rationale in the MR.

The Python compiler is able to recognise `if True` to always take the if-side branch and removes the if, and the whole else-side, from the code:

```
In [4]: def f():
   ...:     if True:
   ...:         return 10
   ...:     else:
   ...:         return 20
   ...: 

In [5]: dis.dis(f)
  2           0 NOP

  3           2 LOAD_CONST               2 (10)
              4 RETURN_VALUE
```

this doesn't happen using a generic expression:

```
In [6]: ASYNC=True

In [7]: def f():
   ...:     if ASYNC:
   ...:         return 10
   ...:     else:
   ...:         return 20
   ...: 

In [8]: dis.dis(f)
  2           0 LOAD_GLOBAL              0 (ASYNC)
              2 POP_JUMP_IF_FALSE        4 (to 8)

  3           4 LOAD_CONST               1 (10)
              6 RETURN_VALUE

  5     >>    8 LOAD_CONST               2 (20)
             10 RETURN_VALUE
```

Not only this works as an optimization, Mypy is also clever enough to apply the same consideration, so it doesn't complain if the else-side contains code that is invalid in the async context, but will become valid in the sync one (such as wrong return type, missing import because they are if'ed away too, etc).

However, I have just noticed that a constant string works too. Maybe it would be more readable to use `if "async"` or `if "ASYNC"` (any case) as a test?

dvarrazzo replied:
I've checked and, while Python is smart enough to prune branches under `if "async"` and `if not "async"`, MyPy is not so clever and raises error as if those branches were reachable.

We might raise a bug to MyPy to ask for more complete constant folding, but for the moment I think this solution is as good as it gets.

dvarrazzo replied:
I have found a related feature request in MyPy and added our use case: https://github.com/python/mypy/issues/12583

dvarrazzo commented on psycopg_pool/psycopg_pool/_task.py:
```diff
@@ -0,0 +1,27 @@
+from typing import Any, Callable, Optional, NamedTuple
+
+
+class Task(NamedTuple):
+    time: float
+    action: Optional[Callable[[], Any]]
+
+    def __eq__(self, other: "Task") -> Any:  # type: ignore[override]
+        return self.time == other.time
+
+    def __lt__(self, other: "Task") -> Any:  # type: ignore[override]
+        return self.time < other.time
+
+    def __le__(self, other: "Task") -> Any:  # type: ignore[override]
+        return self.time <= other.time
+
+    def __gt__(self, other: "Task") -> Any:  # type: ignore[override]
+        return self.time > other.time
+
+    def __ge__(self, other: "Task") -> Any:  # type: ignore[override]
+        return self.time >= other.time
```
I can test that. Code started with Python 3.6 compatibility but master supports Python 3.8 only. 3.1 supports Python 3.7 only, but probably it's not worth backporting and risking something.

dvarrazzo replied:
That was almost right, thank you :slightly_smiling_face:  It just needed `@dataclass(order=True)`.

dvarrazzo commented on tools/convert_async_to_sync.sh:
```diff
@@ -0,0 +1,77 @@
+# Convert all the auto-generated sync files from their async counterparts.
+
+set -euo pipefail
+# set -x
```
Yes, can probably remove it. It was generated by [an editor snippet of mine](https://github.com/dvarrazzo/dotfiles/blob/fcbc13e9ac0fcd320426f81d3252cdb9a5f2a8c5/vim/UltiSnips/sh_piro.snippets#L6-L10) as a reminder to enable echo mode for debugging, but I guess that, without a comment, it works as a remainder for me only (and, after a few years putting bash together, I don't need further reminder).

I'll remove it and look for other occurrences too.

dvarrazzo commented on tools/async_to_sync.py:
```diff
@@ -0,0 +1,462 @@
+
+class AsyncToSync(ast.NodeTransformer):
+    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef) -> ast.AST:
+        new_node = ast.FunctionDef(
+            name=node.name,
+            args=node.args,
+            body=node.body,
+            decorator_list=node.decorator_list,
+            returns=node.returns,
+        )
```
I'll check if it works, or if it fails because there is unexpected stuff in the constructor, thank you

dvarrazzo replied:
Tested, it works, no difference in the output :slightly_smiling_face:

dvarrazzo commented on psycopg_pool/psycopg_pool/sched_async.py:
```diff
@@ -0,0 +1,89 @@
+"""
+A minimal scheduler to schedule tasks run in the future.
```
I will apply the various docstrings/comments suggestions together; also want to double-check if the async->sync conversion doesn't leave incorrect info. Can probably use a pattern such as:

```
"""
Objects used to frobnicate (async version).
"""
```

and convert every `(async` patterns to `(sync`.

dvarrazzo commented:
> One concern is that if we're auto-generating both code and the associated tests, is there something to catch the scenario where we fail to do so? (e.g. by accidentally not adding to the hard-coded list of targets in convert_async_to_sync.sh) I guess this will be easily visible in the diffs, but for a large-scale change like this it seems hard to be sure that nothing has gone missing somewhere?

I have [added a self-check](https://github.com/psycopg/psycopg/pull/657/commits/4f542fd2c79f0169eeb402987955e8148c0e1bb0) to `--check` to verify that all the files declared as generated by the conversion script are actually converted by it.

asqui commented:
Nice! Very fancy bash-fu!

dlax reviewed (commented):
The changeset is too long for me to be read carefully, I appreciate that commits are atomic in general though, but the overall is just too long. So I'm mostly mentioning superficial things I could notice or raising more high-level questions (some might be non-sensical).

I don't quite get the *need* for this infrastructure change, nor do I know if this will be beneficial or not in the long run. Is the maintenance of the sync+async code such a burden? Can you detail a bit more your motivation @dvarrazzo? The approach seems quite unusual to me as well, is there something other projects have been doing?

That being said, changes look okay to me; I would certainly need some time to get used to the new workflow if/when I'd hack back on the code base but will find probably my way through it though.

But since you've done all the work upfront and are presenting this with a done deal, I'm not sure there is place for a lot of discussion. Honestly, this makes me a bit sad as I feel (fear?) this project will ever stay as a one-person effort...

Still some generic questions:

* Alternatively to generating test files, have you considered using pytest fixtures to make the test suite parametrized? Maybe this would be hard to do because of the `async def` and `await ...` syntax that seems hard to parametrized, so just asking.
* At some point, the `convert_async_to_sync.sh` is getting fairly complex (way beyond my limit for shell scripting), have you considered writing this in Python?
* Have you considered following the [sans-io][] pattern? I have no experience in this myself, and I wonder if this is even possible considering we have the libpq in between, which effectively does I/O; so just asking as well.

[sans-io]: https://sans-io.readthedocs.io/

dvarrazzo commented:
Hello @dlax 

> But since you've done all the work upfront and are presenting this with a done deal, I'm not sure there is place for a lot of discussion. Honestly, this makes me a bit sad as I feel (fear?) this project will ever stay as a one-person effort...

First of all, I sincerely apologise for your bad feeling. Especially given to the person who has contributed the most to the project.

It wasn't my intention to present this as a done deal: if you have an opinion for which you consider not a good idea to do this operation I am very happy to consider it.

> I don't quite get the need for this infrastructure change, nor do I know if this will be beneficial or not in the long run. Is the maintenance of the sync+async code such a burden? Can you detail a bit more your motivation @dvarrazzo? The approach seems quite unusual to me as well, is there something other projects have been doing?

The sync+async code maintenance was getting a bit burdensome, yes. Most changes needed to be done on one side required to be reflected on the other side, and the code is difficult to diff. During the conversion I found a couple of little bugs ([this](https://github.com/psycopg/psycopg/commit/8d10d4b1931f9dda8b38ffd662165058c89d0169) and [this](https://github.com/psycopg/psycopg/commit/0c0972957bb3886cd0ab249e81064c44011f1ca3)) and I am actually surprised that there aren't more, which means that we have actually been very careful in writing the code... but it's still a somewhat present danger. There were more differences in the tests, but that's less than an issue.

The branch started mostly as an experiment: I considered that the async side is more or less just the sync side with some more information, and I wanted to see if it was possible to strip these info using AST transformation. Fast forward more hacking and it turned out it was.

I am not aware of other projects following the same approach, as I don't know of others who present a set of sync/async objects pair trying to keep the same semantics. Most projects usually take only one of the sides.

Transforming the code and committing it seems the less magic option: any change introduced would be visible via git diff. The approach followed was actually pretty much to go file-by-file and working either via simple refactoring (e.g. moving the base classes from the sync side to a separate module) and modifying the sync-to-async script until the changes on the sync side were negligible and no meaningful changes were introduced in the already transformed code.

> The changeset is too long for me to be read carefully, I appreciate that commits are atomic in general though, but the overall is just too long. So I'm mostly mentioning superficial things I could notice or raising more high-level questions (some might be non-sensical).

Yes, that's why I didn't ask for a review, but more a feedback about the approach. In terms of diff line count it's massive, but it's mostly about separating the code that can be converted async -> sync from the code that don't need or cannot be converted.

> That being said, changes look okay to me; I would certainly need some time to get used to the new workflow if/when I'd hack back on the code base but will find probably my way through it though.

> Still some generic questions:

> * Alternatively to generating test files, have you considered using pytest fixtures to make the test suite parametrized? Maybe this would be hard to do because of the async def and await ... syntax that seems hard to parametrized, so just asking.

Yes, I don't see a way to do that. Before starting this operation I have actually further reduced the number of differences by parametrising a set of tests that were actually copypasted (Null vs. NonNull pools, normal, Client, Raw cursors). That refactoring is outside this branch (I think) and it was positive in reducing the number of tests written. But unifying sync+async doesn't seem doable at fixture level. 

> * At some point, the convert_async_to_sync.sh is getting fairly complex (way beyond my limit for shell scripting), have you considered writing this in Python?

Yes, that can be done. Especially if we want to make `--check` not modifying the code; that might start getting convoluted in bash.

> * Have you considered following the [sans-io](https://sans-io.readthedocs.io/) pattern? I have no experience in this myself, and I wonder if this is even possible considering we have the libpq in between, which effectively does I/O; so just asking as well.

I read of sans-io and thought about that approach. With the responsibilities we leave to the libpq in the middle it didn't seem obvious how to apply it.

dvarrazzo commented on psycopg_pool/psycopg_pool/null_pool.py:
```diff
@@ -190,13 +176,10 @@ def resize(self, min_size: int, max_size: Optional[int] = None) -> None:
         Only *max_size* can be changed; *min_size* must remain 0.
         """
-        min_size, max_size = self._check_size(min_size, max_size)
+        (min_size, max_size) = self._check_size(min_size, max_size)
```
Uhm... this makes me think that there are differences in what the Python ast module generates.

On my machine, and [in CI](https://github.com/psycopg/psycopg/blob/dcecab35184c591ba84a409157995b0a5ba2349d/.github/workflows/lint.yml#L25), Python 3.10 is used. What Python version are you using?

dlax replied:
I'm using Python 3.11.

dvarrazzo replied:
Yes, that's confirmed: running:

```sh
$ docker run --rm -ti -v `pwd`:/src --workdir /src python:3.11 bash
# pip install ./psycopg[dev]
# ./tools/convert_async_to_sync.sh 
```

I get the same diffs of yours.

dvarrazzo replied:
I have added an option `--docker` to the script. Using this option, the script builds a docker image using Python 3.11 and runs itself in the image. This allows devs to run the script also when the Python version used for the conversion is not available on their machine.

dvarrazzo commented on tools/convert_async_to_sync.sh:
```diff
@@ -73,4 +73,24 @@ if [[ $check ]]; then
+    # Note: the 'cd' early in the script ensures that cwd is the project root.
+    checked=0
+    errors=0
+    for fn in $(find psycopg psycopg_pool tests -type f -name \*.py); do
+        # Skip files ignored by git (build artifacts)
+        ! git status --porcelain --ignored "${fn}" | grep -q '!!' || continue
```
Can consider that (especially if we rewrite this script in Python, or we extend `async_to_sync.py` to add the options to work on all the project's files.

dvarrazzo replied:
Using `git ls-files` in the python porting of the bash script.

https://github.com/psycopg/psycopg/blob/6f49a9b727b89fe0a78b244a45f3798d9eea0a21/tools/async_to_sync.py#L95C9-L95C9

dvarrazzo commented on psycopg_pool/psycopg_pool/_acompat.py:
```diff
@@ -0,0 +1,47 @@
+
+class AEvent(asyncio.Event):
+    """
+    Subclass of asyncio.Event adding a wait with timeout like threading.Event.
+
```
Uhm... actually maybe we could transform `asyncio.wait_for(ev.wait(), timeout=n)` to `ev.wait(timeout=n)`. We don't have, at ast level, really information about the type of `ev` but it is not really used in so many places and with so many different objects that we can use a convention on the name (to distinguish events from queues from conditions...)

Sounds fun. And - apologising again for presenting this work as a done thing - pretty much being fun, with the added benefit of reducing the amount of code to actively maintain and uniforming async/sync behavioural differences, was the main motivation of this work :sweat_smile:

dlax replied:
Looking more at httpcore, they handle this case by wrapping the underlying `Event` (from `threading` or `asyncio`) into their own `Event` class, which thus has a `def wait(timeout=):`; maybe we could follow this route?

https://github.com/encode/httpcore/blob/8780c9cf11ccd5d0a5ddc307bd2b1ed8a682a030/httpcore/_synchronization.py

dvarrazzo commented on tests/_test_cursor.py:
```diff
@@ -0,0 +1,60 @@
+        return query
+
+    if "%(" in query:
+        raise pytest.skip("RawCursor only supports positional placeholders")
```
You are right, it's a mistake. It's a pattern slipped from psycopg2 unittest-based test suite (possibly for compatibility with the unittest shipped with ancient Python versions)

dlax commented:
> I am not aware of other projects following the same approach, as I don't know of others who present a set of sync/async objects pair trying to keep the same semantics. Most projects usually take only one of the sides.

Just discovered that [httpcore](https://github.com/encode/httpcore/) does the same, see [unasync.py](https://github.com/encode/httpcore/blob/8780c9cf11ccd5d0a5ddc307bd2b1ed8a682a030/unasync.py), though with a different approach (regex). Then, it's "frontend", httpx, apparently keeps both sync/async interfaces hand-written.

dvarrazzo commented:
@dlax 

> At some point, the convert_async_to_sync.sh is getting fairly complex (way beyond my limit for shell scripting), have you considered writing this in Python?

I have dropped the bash script and moved the same features to `async_to_sync.py`: iterate on the list of known files, running black on them, check for inconsistencies.

dlax commented:
> I have dropped the bash script and moved the same features to `async_to_sync.py`: iterate on the list of known files, running black on them, check for inconsistencies.

Thank you! I've re-read the script, it looks okay to me. Just prepared a few improvements in #660 while reading.

dvarrazzo commented:
Merged! Thank you very much for your feedback, @dlax, @asqui!
